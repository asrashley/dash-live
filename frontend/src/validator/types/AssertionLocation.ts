export interface AssertionLocation {
    filename: string;
    line: number;
}
