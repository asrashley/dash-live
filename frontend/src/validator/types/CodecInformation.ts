export interface CodecDetails {
    label: string;
    error?: string;
    details: string[];
}

export interface CodecInformation {
    codec: string;
    error?: string;
    details: CodecDetails[];
}
