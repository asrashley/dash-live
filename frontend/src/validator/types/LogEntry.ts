export type LogEntry = {
    id: number;
    level: 'error' | 'info' | 'debug';
    text: string;
};
