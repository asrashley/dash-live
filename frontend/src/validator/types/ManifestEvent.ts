
export interface ManifestEvent {
    text: string[];
}