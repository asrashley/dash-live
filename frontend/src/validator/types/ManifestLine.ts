export interface ManifestLine {
    line: number;
    text: string;
    hasError: boolean;
    errors: string[];
}
