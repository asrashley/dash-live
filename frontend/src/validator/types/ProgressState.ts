export interface ProgressState {
    currentValue?: number;
    minValue: number;
    maxValue: number;
    text: string;
    finished?: boolean;
    error?: boolean;
 }
