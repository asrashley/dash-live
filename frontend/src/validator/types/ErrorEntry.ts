import { AssertionLocation } from "./AssertionLocation";

export interface ErrorEntry {
    assertion: AssertionLocation;
    location: [number, number];
    clause?: string;
    msg: string;
}