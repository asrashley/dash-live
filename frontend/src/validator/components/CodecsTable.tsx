import { useComputed, type ReadonlySignal } from "@preact/signals";

import { CodecDetails, CodecInformation } from "../types/CodecInformation";

function CodecDetailsCell({ details, error, label }: CodecDetails) {
  if (error) {
    return <p className="codec-error fw-bold">{error}</p>;
  }
  return (
    <div>
      <p class="codec-item fw-bold">{label}</p>
      {details.map((item) => (
        <p key={item} className="codec-item">{item}</p>
      ))}
    </div>
  );
}

function CodecRow({ codec, details }: CodecInformation) {
  return (
    <tr>
      <td className="codec-string">{codec}</td>
      <td className="codec-detail">
        {details.map((det) => (
          <CodecDetailsCell key={det.label} {...det} />
        ))}
      </td>
    </tr>
  );
}

interface CodecsTableProps {
  codecs: ReadonlySignal<CodecInformation[]>;
}

export function CodecsTable({ codecs }: CodecsTableProps) {
  const className = useComputed<string>(() =>
    codecs.value.length === 0 ? "d-none" : "table table-striped manifest-codecs"
  );
  return (
    <table className={className}>
      <thead>
        <tr>
          <th class="codec-string">Codec</th>
          <th class="codec-detail">Details</th>
        </tr>
      </thead>
      <tbody>
        {codecs.value.map((info) => (
          <CodecRow key={info.codec} {...info} />
        ))}
      </tbody>
    </table>
  );
}
