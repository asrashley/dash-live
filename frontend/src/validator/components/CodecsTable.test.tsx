import { signal } from "@preact/signals";
import { beforeEach, describe, expect, test } from "vitest";
import { CodecInformation } from "../types/CodecInformation";
import { renderWithProviders } from "../../test/renderWithProviders";
import { CodecsTable } from "./CodecsTable";
import { exampleCodecs } from "../../test/fixtures/exampleCodecs";

describe("CodecsTable component", () => {
  const codecs = signal<CodecInformation[]>([]);

  beforeEach(() => {
    codecs.value = structuredClone(exampleCodecs);
  });

  test('matches snapshot', () => {
    const { asFragment, getByText, getBySelector } = renderWithProviders(<CodecsTable codecs={codecs} />);
    codecs.value.forEach(item => {
        getByText(item.codec);
        item.details.forEach(det => {
            det.details.forEach(line => getByText(line));
        });
    });
    expect(getBySelector('table').classList.contains('d-none')).toEqual(false);
    expect(asFragment()).toMatchSnapshot();
  });

  test('displays error', () => {
    codecs.value[0].details[0].error = 'an error message';
    const { getByText } = renderWithProviders(<CodecsTable codecs={codecs} />);
    getByText('an error message');
  });

  test('hides table when there are no codecs', () => {
    codecs.value = [];
    const { getBySelector } = renderWithProviders(<CodecsTable codecs={codecs} />);
    expect(getBySelector('table').classList.contains('d-none')).toEqual(true);
  });
});
