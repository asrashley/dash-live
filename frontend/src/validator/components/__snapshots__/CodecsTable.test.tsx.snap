// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodecsTable component > matches snapshot 1`] = `
<DocumentFragment>
  <table
    class="table table-striped manifest-codecs"
  >
    <thead>
      <tr>
        <th
          class="codec-string"
        >
          Codec
        </th>
        <th
          class="codec-detail"
        >
          Details
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="codec-string"
        >
          avc3.640028
        </td>
        <td
          class="codec-detail"
        >
          <div>
            <p
              class="codec-item fw-bold"
            >
              AVC/H.264
            </p>
            <p
              class="codec-item"
            >
              profile_idc=100 constraint_set=0 level_idc=40
            </p>
            <p
              class="codec-item"
            >
              profile=High (64)
            </p>
            <p
              class="codec-item"
            >
              constraints=------
            </p>
            <p
              class="codec-item"
            >
              level=4 (28)
            </p>
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="codec-string"
        >
          mp4a.40.2
        </td>
        <td
          class="codec-detail"
        >
          <div>
            <p
              class="codec-item fw-bold"
            >
              AAC
            </p>
            <p
              class="codec-item"
            >
              MPEG-4 AAC (40)
            </p>
            <p
              class="codec-item"
            >
              Low-Complexity AAC (2)
            </p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
